### ISSUE TITLE
Graphics tablet support

### ISSUE THREAD

**Issue by jp7677** [labwc/labwc#1060]
Hi there,

I'm using a Wacom tablet (Intous M) most of the time as a mouse replacement. The tablet works perfectly when using Sway, but nothing happens in labwc when moving the stylus, tapping the surface or clicking a button (a normal mouse works though). Since it works just fine in Sway, I guess `wlroots` has tablet support, but it is just not (yet) implemented in labwc? If that is the case, could you please consider adding support for tablets in labwc?

Kind regards and thanks a lot for your cool project here!

**Comment by Consolatis** [comment]
> I guess `wlroots` has tablet support, but it is just not (yet) implemented in labwc?

This is correct.

I've played around with it quite a while ago to figure out if we could theoretically support drawing tablets and how much work it would be but stopped once seeing the events come in. Integrating the drawing tablet input into our cursor handling might be a bit more complex. I'll try to find my old test branch and look into it again but can't give any promises.

**Comment by jp7677** [comment]
Hi @Consolatis , out of curiosity, did you found your old branch and is it somewhere publicly available?

**Comment by jp7677** [comment]
Thanks a lot for your response and for your work in this direction. Fingers crossed for finding your power supply!

>  I'll prioritize this over other enhancements for now. But fixing bugs is still the main priority. So I can't provide any kind of time frame for this.

Very cool and yes, I totally understand that this is a niche feature and finding and fixing bugs has priority.

**Comment by Consolatis** [comment]
> Fingers crossed for finding your power supply!

Well, on the one side that does appear to have helped, found that power supply.
On the other side the branch wasn't there either. So I started fresh and pushed a first very basic initial prototype, feel free to test: https://github.com/Consolatis/labwc/tree/feature/drawing_tablet_support

Some caveats:
- only supports mouse emulation for "tools" / pens
- buttons are not supported, "tip down" is treated as left mouse button press, "tip up" as release
- does not support the wayland tablet protocol. So applications that interact with things like pressure or tilting won't work and will just see a usual mouse
- the pad itself is not supported, so if it contains further things like buttons or wheels or anything fancy like that those won't work
- it will spam the log, unless you want to debug the events I recommend not running labwc with the `-V` or `-d` flags while using this branch
- no guarantees are made about the stability. I tested it a bit and didn't spot anything in that regard but it really is just a first prototype to figure out how to best hook that up into the rest of labwc

**Comment by jp7677** [comment]
Yeah, just testing your experimental branch and it works already very nicely. Thanks a lot for starting on this! Please let me know when the time comes for detailed feedback, I would love to help you with getting this into shape!

Personally I think from the features of your list, once button mapping, output mapping and aspect ratio can be configured, it already becomes very much usable as a mouse replacement, which is my use case.

**Comment by Consolatis** [comment]
I am not entirely sure how the buttons could be mapped, every pen might have different ones. My cheap test setup for example has two buttons with one emitting `BTN_TOOL_BRUSH` and the other `BTN_TOOL_PENCIL`. I could just always pretend they are `BTN_LEFT` (for the mouse emulation) and things should just work. However, that would also mean that with a pen there would be no way to use other buttons (again, for the mouse emulation part). Another option would be not to translate them at all and require the user to define every config entry again for use with a graphics tablet (which I think would be really annoying if you look at [docs/rc.xml.all](https://github.com/labwc/labwc/blob/master/docs/rc.xml.all#L235) and it wouldn't work without a user config).

Personally I never use the tablet so I also don't know if for example the `tip down/up` events should even emulate a mouse button or not. I could simply ignore the tip for the mouse emulation and only forward the buttons instead.

For the output mapping it is also not that easy to figure out how to present a user facing configuration properly. The actual mapping should be relatively trivial though.

Similar issue for adjusting the position of the pad, (e.g. switching the x and y axis or inverting one or both). It would be relatively trivial to do code wise but I am not sure how we could provide a proper config for that.

**Comment by jp7677** [comment]
Very cool, all three buttons on the stylus are working now. Also the default mapping without any configuration is perfect for me.

Regarding the output mapping, aspect ratio and  rotation, what about just adding those as attributes to the `tablet` element, something like this?

```xml
<tablet mapToOutput="HDMI-A-1" keepAspectRatio="true" orientation="landscape">
	<map button="tip" to="left" />
	<map button="stylus" to="right" />
	<map button="stylus2" to="middle" />
</tablet>
``` 

I would keep it simple and just bind to the primary monitor by default unless configured otherwise. Not sure if mapping to multiple outputs is a common scenario that would justify a much more complex configuration and implementation.

**Comment by Consolatis** [comment]
> Regarding the output mapping, aspect ratio and rotation, what about just adding those as attributes to the `tablet` element, something like this?

In general I like this. `mapToOutput` seems good. not sure about the other two
- I don't quite understand what `keepAspectRatio` is supposed to do
- In regards to `orientation` we'd need to have 4 options here, e.g. `landscape` and `landscape-inverted`. If something being landscape or not would depend on the specific tablet. So I tend to prefer something like `invert_axis="x"` / `invert_axis="y"` or `invert_axis="xy"` instead.

> I would keep it simple and just bind to the primary monitor by default unless configured otherwise.

Funnily enough, mapping the input to specific outputs is actually more difficult than to span all outputs.
I'd think this also depends on the tablet, e.g. if it has an own screen or not. We could also think about having an action that allows changing the output mapping during runtime. That could then be bound on a pad or tool button.
All of this is for the mouse emulation part only though. Once implemented, an application using the tablet protocol directly would just have its own region mapped to the tablet.

> Not sure if mapping to multiple outputs is a common scenario that would justify a much more complex configuration and implementation.

See previous point, its actually easier to map to all outputs rather than to a single one.
Also, when used for mouse emulation, how would you control windows on a different output then?

> Edit: I also have four buttons on the tablet itself, but they don't even show up in `wev`, so that is probably a very different thing. Would be cool though to use those in or similar like `keyboard/keybind` to bind them to an action. But one step at a time.

Yeah, the pad isn't handled at all so far.

Thanks for your feedback regarding the whole thing btw. Although I do have a cheap test setup I am really not using it at all. So getting feedback from somebody actually using a tablet for mouse emulation is really valuable.

**Comment by jp7677** [comment]
> I don't quite understand what `keepAspectRatio` is supposed to do

My tablet has a 16:10 surface area, whereas my monitor has 21:9 aspect ratio. By keeping the aspect ratio when mapping the desktop to the tablet, the active surface area should be truncated so that the aspect ratio is the same as the one from the desktop. Thus essentially "black boarders" on top/bottom or left/right on the tablet surface so that the aspect ratio matches the screen.
 
> In regards to `orientation` we'd need to have 4 options here, e.g. `landscape` and `landscape-inverted`. If something being landscape or not would depend on the specific tablet. So I tend to prefer something like `invert_axis="x"` / `invert_axis="y"` or `invert_axis="xy"` instead.
 
Good point, what about `orientation` in fixed degrees, thus `0`, `90`, `180` and `240`? Assuming I understood your proposal correctly, inverting the axes would flip/mirror the coordinates, I think this is a different transformation. 
 
> Funnily enough, mapping the input to specific outputs is actually more difficult than to span all outputs. I'd think this also depends on the tablet, e.g. if it has an own screen or not. We could also think about having an action that allows changing the output mapping during runtime. That could then be bound on a pad or tool button. All of this is for the mouse emulation part only though. Once implemented, an application using the tablet protocol directly would just have its own region mapped to the tablet.

Sounds good. I thought that having a monitor configuration with different resolutions and e.g. two next to each other, may be rotated, and one on top  (thus forming everything except a rectangle) would make things difficult. Cool to hear that I'm wrong ;) 

> Thanks for your feedback regarding the whole thing btw. Although I do have a cheap test setup I am really not using it at all. So getting feedback from somebody actually using a tablet for mouse emulation is really valuable.

You are very welcome and I love to continue giving feedback! This is also all in my interest since I'm really would like to see good tablet support (that is primary mouse emulation, I'm no artist ;)). 
I switched from mouse to tablet a few years ago due to RSI effects on my arms. The more natural position of arms (not that twisted anymore) works very well for me and kept me in the business. Unfortunately the tablet support on Gnome X11 was perfect but is unusable on Gnome Wayland due lots of smaller issues. While looking for alternatives I found labwc and once I had a configuration for my needs and a good integration setup (waybar and friends) I figured this might be the way to go for me (also considering that sway as another wlroots compositor has a pretty good tablet support). So I really appreciate  the work you and other maintainers are doing here including your openness and responsiveness to feedback!

**Comment by Consolatis** [comment]
> My tablet has a 16:10 surface area, whereas my monitor has 21:9 aspect ratio. By keeping the aspect ratio when mapping the desktop to the tablet, the active surface area should be truncated so that the aspect ratio is the same as the one from the desktop. Thus essentially "black boarders" on top/bottom or left/right on the tablet surface so that the aspect ratio matches the screen.

Hm.. not sure about that feature. Especially for smaller tablets having even less space seems not that desirable.
Also I'd assume in most cases the tablet does not have the exact aspect ratio as the outputs. Certainly not for most multi output layouts. Thus this function would just waste space on the tablet and complicate the code quite a bit.



> Good point, what about using `rotateBy` with fixed degrees, thus `0`, `90`, `180` and `240`? Assuming I understood your proposal correctly, inverting the axes would flip/mirror the coordinates, I think this is a different transformation than what I had in mind.

Yep, `rotate="0|90|180|270"` with default to `0` is *much* better. Thanks!

> Sounds good. I thought that having a monitor configuration with different resolutions and e.g. two next to each other, may be rotated, and one on top (thus forming everything except a rectangle) would make things difficult. Cool to hear that I'm wrong ;)

That should be handled by our cursor backend already as its the same situation for usual mouse usage.

**Comment by jp7677** [comment]
> Hm.. not sure about that feature. Especially for smaller tablets having even less space seems not that desirable. Also I'd assume in most cases the tablet does not have the exact aspect ratio as the outputs. Certainly not for most multi output layouts. Thus this function would just waste space on the tablet and complicate the code quite a bit. I might look into it after everything else is done but I think this is a pretty low priority feature.

Yeah it is a trade-off, sure. Personally I prefer to loose some tablet space, but have a consistent movement of the stylus. If I turn off that feature, the effect is that when moving the stylus 1cm to the bottom, the cursor moves 1cm down on the screen, but when moving the stylus 1cm to the right, the cursor moves 2cm to the right on the screen.  You can get used to it, but it doesn't feel natural. Agree though that all the basics and rotation are more important, I would also keep it off by default.
 
> Yep, `rotate="0|90|180|270"` with default to `0` is _much_ better. Thanks!

My math left me, it's of course `270` instead of `240`, thanks ;)

**Comment by Consolatis** [comment]
Small update in the experimental branch:
- change to the new config format, see example below
- add support for `rotate`
- allow usual mousebinds (including modifiers) for the new buttons (`tip`, `stylus`, `stylus2`, `stylus3`). By default, the first three are still mapped to the left, right and middle mouse buttons though so currently to use a bind for e.g. `stylus2` it has to be manually mapped to itself again to overwrite the default mapping. Need to think about how to combine this feature with the default mouse button emulation so mapping to itself isn't necessary anymore in the future.

```xml
<tablet rotate="90">
  <map button="stylus2" to="stylus2" />
</tablet>
<mouse>
  <default />
  <context name="Root">
    <mousebind button="W-Stylus2" action="Click">
      <action name="Execute" command="xeyes" />
    </mousebind>
  </context>
```

This example config should make `xeyes` show up when holding the logo button on the keyboard and pressing the 2nd pen button on the empty desktop. Next on the list when working on this again: output mapping.

**Comment by jp7677** [comment]
While playing a bit with your branch I've observed the following effects. I know of course that this is early stage, but may it is useful to already have some detailed feedback. I can imagine that a lot of points have the same root cause, so it is probably less than it appears. I also hope the points are readable, sometimes I found it difficult to find the right words and to recognize the correct circumstances (toolkit, decoration etc.). All those actions and effects I'm describing do work correctly when being performed with a normal mouse.  Also of course very cool progress on the stylus/mouse binding! This offers great flexibility. 

### Hovering with the stylus

When hovering with the stylus very (very, very) slowly, it sometimes feels like the cursor is lagging behind for a second and then jumps to the new position. No big deal though and took me a while to even realize this.  

### Resizing application windows

- Firefox-Wayland (gtk3)

The resize-window cursor appears on hovering at window border, but stays like this when hovering into the window. The cursor changes into the normal shape one only after a stylus tap, it should change as soon as hovering away from the border. Resizing itself does work with the stylus. 

- Tilix (gtk3, CSD)

The resize-window cursor never appears when hovering on the border, also resizing itself is not possible.

- Gnome-Terminal (gtk4, SSD)

Resizing cursor appearance and resizing the window does work fine. 

### Moving application window by grabbing the title bar

- Firefox-Wayland (gtk3), Tilix (gtk3, CSD)

Grabbing the title bar and moving the window works, but only happens after stylus tap release. It should work immediately after grabbing with stylus tap without stylus tap release.  

- Gnome-Terminal (gtk4, SSD)

Moving the window does work fine, also immediately on stylus tap.

### Text input

- Firefox-Wayland (gtk3), Tilix (gtk3), Gnome-Terminal (gtk4, SSD) 

When hovering over a text input area, the cursor shape does not change. A stylus tap is needed to change the cursor into the edit-shape. The shape should change when only hovering over a text input area without a tap.  
When entering text with the keyboard, the edit cursor correctly disappears. When starting to hover the stylus again the cursor stays hidden. A stylus tap or hover outside of the window is needed to make the cursor visible again. The cursor should also become visible when only hovering over the application window without a stylus tap.

### Selecting text

- Firefox-Wayland (gtk3), Tilix (gtk3), Gnome-Terminal (gtk4, SSD)

Selecting text using the stylus in one of those applications does work, but the text selection becomes visible only after the stylus tap release. The selected text should "grow" when holding the tap and moving the stylus over the text.
A similar effect (result visible only after stylus tap release) can be seen on using scrollbars with the stylus.

### No stylus tap interaction or interaction on wrong position

- Spotify (xwayland, electron, flatpak), VS-Code (xwayland, electon, flatpak), IntelliJ (xwayland)

When using one of those applications I cannot tap/click at all or the tap is out of position, like the tap/click happens on the wrong place. May be xwayland is the issue here?

**Comment by Consolatis** [comment]
Thanks for the detailed feedback. Will look into it the next time I am working on the feature.

**Comment by jp7677** [comment]
> Edit: Random question, do you use output scaling?

No, just a single monitor with no output scaling.

**Comment by jp7677** [comment]
I think I fixed all of the cursor issues I mentioned above with this addition on top of your branch:

```diff
diff --git a/src/input/cursor.c b/src/input/cursor.c
index 32b7a08..d4f8bb6 100644
--- a/src/input/cursor.c
+++ b/src/input/cursor.c
@@ -1035,6 +1035,7 @@ cursor_emulate_move_absolute(struct seat *seat, double x, double y,
 
        wlr_cursor_move(seat->cursor, NULL, dx, dy);
        process_cursor_motion(seat->server, time_msec);
+       wlr_seat_pointer_notify_frame(seat->seat);
 }
 
 void
```

Running `wev` I saw regular `frame` events when move the mouse, but no frame events with stylus. Not sure if this is the correct place, but something like this should imho do the trick.

**Comment by Consolatis** [comment]
Nice catch! This totally makes sense. Will add.

**Comment by jp7677** [comment]
> When hovering with the stylus very (very, very) slowly, it sometimes feels like the cursor is lagging behind for a second and then jumps to the new position. No big deal though and took me a while to even realize this.

I could track this down as well. Those "lags" happen when only the x or the y axis is updated. The current code drops events when either `x` or `y` is `0.0`. Unfortunately that is the case when moving the stylus in a straight vertical or horizontal line, thus when only one axis is updated. The effect was that the cursor didn't move until my hand wasn't steady enough anymore  and I had a change in the other dimension as well. 
I could work around this by adding `double x, y;` as "last known coordinates" in `struct drawing_tablet`, initialize those in `setup_pen` and change the event handler into this:

```c
if (ev->updated_axes & (WLR_TABLET_TOOL_AXIS_X | WLR_TABLET_TOOL_AXIS_Y)) {
    wlr_log(WLR_DEBUG, "Got new axis pos: %.10f,%.10f", ev->x, ev->y);
    if (ev->x > 0.0 || ev->y > 0.0) {
        if (ev->x > 0.0) {
            tablet->x = ev->x;
        }
        if (ev->y > 0.0) {
            tablet->y = ev->y;
        }

        double x = tablet->x;
        double y = tablet->y;
        adjust_for_aspect_ratio(tablet->tablet->width_mm, tablet->tablet->height_mm, &y);
        adjust_for_rotation(rc.tablet.rotate, &x, &y);
        cursor_emulate_move_absolute(tablet->seat, x, y,
                &ev->tablet->base, ev->time_msec);
    } else {
        wlr_log(WLR_DEBUG, "Blocking strange axis pos: %.10f, %.10f", ev->x, ev->y);
    }
}
```

(There should be an even easier way since the last cursor position is already available in `seat->cursor`, but somehow  `tablet->seat->cursor`  didn't work)

Now the cursor movement with the stylus is butter smooth and overall the tablet works perfectly in my local `labwc` build.

**Comment by Consolatis** [comment]
Hm.. I assume rather than checking for `> 0.0` we should instead just check if the axis was even updated (e.g. `ev->updated_axes & WLR_TABLET_TOOL_AXIS_XXX`. Likely the `0.0` values come from there. That still means that we need the last position though. The result should be the same as with your patch.

Could you post your implementation of `adjust_for_aspect_ratio()` ?
I'll try to work on this a bit more in the next days and also create a draft PR for it, that should make it a bit easier to follow / discuss the code.

**Comment by jp7677** [comment]
> Hm.. I assume rather than checking for `> 0.0` we should instead just check if the axis was even updated (e.g. `ev->updated_axes & WLR_TABLET_TOOL_AXIS_XXX`. Likely the `0.0` values come from there. That still means that we need the last position though. The result should be the same as with your patch.

Yeah, thinking about this, this is much better since `0.0` is of course a valid position.
 
> Could you post your implementation of `adjust_for_aspect_ratio()` 

My simple implementation (which does work) looks like this:

```c
static void
adjust_for_aspect_ratio(double tablet_width, double tablet_height, double *y)
{
    double output_width = 3440.0;
    double output_height = 1440.0;

    if (tablet_width > 0 && tablet_height > 0) {
        double tablet_height_adjusted = tablet_width * output_height / output_width;
        *y = *y / tablet_height_adjusted * tablet_height;
    }
}
```

Obviously `output_width` and `output_height` should come from the attached output, but I couldn't find a way to get to that since (at least on my setup) outputs are initialized after inputs on startup. I figured it would be easier to first wait until you come up with something for output mapping :).

> ? I'll try to work on this a bit more in the next days and also create a draft PR for it, that should make it a bit easier to follow / discuss the code.

Cool!

**Comment by jp7677** [comment]
Hi @Consolatis If you are good with this, I can open a PR with https://github.com/labwc/labwc/compare/master...jp7677:labwc:tablet-support Those are your first commits with some cleanup and the fixes I mentioned above. Those 3 commits don't touch any configuration options yet, so really just the groundwork. I'm using this on top of 0.6.6. for a while with my specific aspect ratio correction on top of it and haven't seen any issue with it.

**Comment by Consolatis** [comment]
Sorry, I am really low on time for FOSS stuff currently.
Yes, I'd be absolutely fine with you sending in a PR :)

**Comment by charbelnicolas** [comment]
Good to see that tablet support is on the way! Thank you

**Comment by johanmalm** [comment]
Closed by #1306

**Comment by Narrat** [comment]
> Closed by #1306

Still open :D

**Comment by johanmalm** [comment]
> > Closed by #1306
> 
> Still open :D

Good catch. We're a bit behind on issue tidy-ups :smile:

**Comment by Consolatis** [comment]
Just to clarify, we currently only support the mouse emulation part of the graphic tablet support (thanks to @jp7677 !), we don't actually support the wayland protocol for wayland applications receiving the raw tablet input.

**Comment by jp7677** [comment]
I've added two issues for the remaining features that came across this issue which aren't implemented yet. That said. the tablet support for me is very much complete, so both are rather "nice to have" imho. (Both are also a bit trickier to implement, at least from my perspective).

Thanks a lot for you initial work on this @Consolatis , it was a pleasure to continue on your branch!

**Comment by jp7677** [comment]
> we don't actually support the wayland protocol for wayland applications receiving the raw tablet input.

Are there any?
